File: LeetCode/1727/cpp/lc1727.cpp
#include <algorithm>
#include <cassert>
#include <vector>

class Solution {
public:
  int largestSubmatrix(std::vector<std::vector<int>>& matrix) {
    const int N = matrix[0].size();
    // Since we must rearrange entire columns, a structure is needed to hold number of 1's
    // in a given column. Note that the matrix is not guaranteed to be square (i.e. M can != N).
    std::vector<int> histogram(N);
    int ans = 0;
    for (const std::vector<int>& row: matrix) {
      for (int i = 0; i < N; i++) {
        histogram[i] = (row[i] == 0) ? 0 : histogram[i] + 1;
      }

      std::vector<int> sorted(histogram);
      std::sort(sorted.begin(), sorted.end());
      for (int i = 0; i < N; i++) {
        ans = std::max(ans, sorted[i] * (N - i));
      }
    }
    return ans;
  }
};

int main() {
  Solution s;
  {
    std::vector<std::vector<int>> matrix = {{0, 0, 1}, {1, 1, 1}, {1, 0, 1}};
    assert(s.largestSubmatrix(matrix) == 4);
  }
  {
    std::vector<std::vector<int>> matrix = {{1, 0, 1, 0, 1}};
    assert(s.largestSubmatrix(matrix) == 3);
  }
  {
    std::vector<std::vector<int>> matrix = {{1, 1, 0}, {1, 0, 1}};
    assert(s.largestSubmatrix(matrix) == 2);
  }
}

File: LeetCode/1727/python/lc1727.py
from typing import List

# Runtime complexity: O(W*H) + O(sort) ~= O(W*H)
# Auxiliary space complexity: O(W)
# Subjective level: medium-
# Solved on: 2026-03-17
class Solution:
    def largestSubmatrix(self, matrix: List[List[int]]) -> int:
        W = len(matrix[0])
        histogram = [0] * W
        ans = 0
        for row in matrix:
            for i in range(W):
                histogram[i] = 0 if row[i] == 0 else histogram[i] + 1

            sortedHistogram = sorted(histogram)
            # histogram.sort()

            for i in range(W):
                ans = max(ans, sortedHistogram[i] * (W-i))
        return ans
